### PR TITLE
[SPARK-42683] Automatically rename conflicting metadata columns

### PR DESCRIPTION
**Pull request by ryan-johnson-databricks**
### What changes were proposed in this pull request?

Today, if a data source already has an output column called `_metadata`, queries cannot access the file-source metadata column that normally carries that name. 

This PR addresses the issue by automatically renaming any metadata column whose name conflicts with an output column. We also introduce a new dataframe method, `metadataColumn`, which mirrors the existing `col` method but (a) only works for metadata columns; and (b) will reliably find metadata columns even if they were renamed due to a name conflict.

### Why are the changes needed?

Today, it's too easy to lose access to metadata columns if the user's table happened to have the wrong column name. This sharp edge limits the utility of metadata columns in general, because the feature doesn't work reliably for all table schemas.

### Does this PR introduce _any_ user-facing change?

Suppose a table defines a `StringType` column called `_metadata`. Today, the metadata column is not accessible, and the example below would return the table's string column:
```scala
df.select("_metadata")
```

With the change, the original query still returns a string, but the file-source metadata column can be found and accessed by invoking `DataSet.metadataColumn`:
```scala
df.withColumn("m", df.metadataColumn("_metadata"))
```

NOTE: Adding a SQL user surface for accessing metadata columns with conflicting names is out of scope for this pull request and can be addressed as future work.

### How was this patch tested?

New unit tests added.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -2714,6 +2726,17 @@ class Dataset[T] private[sql](
...
+  /**
+   * Returns a new Dataset by selecting a metadata column with the given logical name.
+   *
+   * A metadata column can be accessed this way even if the underlying data source defines a data
+   * column with a conflicting name.
+   *
+   * @group untypedrel
+   * @since 4.0.0
+   */
+  def withMetadataColumn(colName: String): DataFrame = withColumn(colName, metadataColumn(colName))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2738]: people can just do `df.withColumn(name, df.metadataColumn(col))`, right?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:2738]: This is a weird API as there is no `withColumn` function that takes only a string parameter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 2]
```diff
@@ -42,6 +42,24 @@ abstract class LogicalPlan
    */
   def metadataOutput: Seq[Attribute] = children.flatMap(_.metadataOutput)
 
+  /**
+   * Finds a metadata attribute of this node by its logical name. This search will work even if the
+   * metadata attribute was renamed because of a conflicting name in the data schema.
+   */
+  def getMetadataAttributeByName(name: String): Attribute = {
+    // NOTE: An already-referenced column might appear in `output` instead of `metadataOutput`.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:50]: metadata col may appear in `output`, but always appear in `metadataOutput`. I think `outputMetadataAttributes.resolve(nameParts, resolver)` should do the work.

**Comment by cloud-fan** [comment]
It's a good idea to provide an API that allows people to unambiguously reference metadata columns, and I like the new `Dataset.metadataColumn` function. However, I think the prepending underscore approach is a bit hacky. It's too implicit and I'd prefer a more explicit syntax like `SELECT metadata(_metadata) FROM t`. We can discuss this more and invite more SQL experts. Shall we exclude it from this PR for now?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 2 above)
Comment (olaky) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:50]: I had an instance in the debugger where this was not the case yesterday

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 3]
```diff
@@ -42,6 +42,24 @@ abstract class LogicalPlan
...
+    (metadataOutput ++ output).collectFirst {
+      case MetadataAttributeWithLogicalName(attr, logicalName)
+          if conf.resolver(name, logicalName) => attr
+    }.getOrElse {
+      val availableMetadataColumns = (metadataOutput ++ output).collect {
+        case MetadataAttributeWithLogicalName(_, logicalName) => logicalName
```
Comment (olaky) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:56]: How about adding MetadataColumn here as well in case we have a metadata column without logical name? Even though this might not be possible right now, it would hurt debuggability if metadata columns are missing from this output

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala [hunk 4]
```diff
@@ -244,6 +245,89 @@ class FileMetadataStructSuite extends QueryTest with SharedSparkSession {
...
+
+    checkAnswer(
+      df.select("_metadata", "__metadata", "___metadata", "____metadata.file_name"),
+      Seq(
+        Row("jack", 24, Row(12345L, "uom"), f0(METADATA_FILE_NAME)),
+        Row("lily", 31, Row(54321L, "ucb"), f1(METADATA_FILE_NAME))
+      )
+    )
+  }
+
```
Comment (olaky) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala:330]: Question about a use case with aliasing: If the metadata column is aliased, we still have to use the original name in withMetadata (or the other functions). So
```
df.select("my_metdata", _metadata).withMetadata("another_metadata_name", "my_metadata")
```
Is not possible I believe? I think this is ok, it just reflects that withMetadata is immune to renaming, but could be good to document this clearly

**Review by olaky — COMMENTED** [review]
Thanks for picking this up, it has been a limitation for too long!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala [hunk 5]
```diff
@@ -340,13 +358,26 @@ trait ExposesMetadataColumns extends LogicalPlan {
...
-      def isOutputColumn(col: AttributeReference): Boolean = {
-        outputNames.exists(name => resolve(col.name, name))
+      // Generate a unique name by prepending underscores.
+      @scala.annotation.tailrec
+      def makeUnique(name: String): String = name match {
+        case name if outputNames.exists(resolve(_, name)) => makeUnique(s"_$name")
+        case name => name
+      }
+
+      // Rename metadata struct columns whose names conflict with output columns.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:368]: Name is not the id of a column, e.g. people can do `SELECT 1 a, 2 a`. I don't quite understand why we need to rename here. What we need is to resolve `_metadata` to the actual metadata `AttributeReference`. At the end Spark will bind `AttributeReference` correctly by looking at `exprId`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:50]: Do we have a test to reproduce it? This should not happen after https://github.com/apache/spark/commit/5705436f70e6c6d5a127db7773d3627c8e3d695a

Comment (ryan-johnson-databricks) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:50]: `SubqueryAlias` does not always propagate its child's `metadataOutput`, even if the child's own `output` includes metadata columns. That can result in a query plan whose `output` contains attributes which satisfy `MetadataAttribute` but are not in `metadataOutput`. 

I raised https://github.com/apache/spark/pull/40321 as a possible fix for that issue.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 5 above)
Comment (ryan-johnson-databricks) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:368]: The immediate issue is that node types implementing `ExposesMetadataColumns` all use its `metadataOutputWithOutConflicts` method to ensure that names metadata columns in `metadataOutput` do not overlap with names of non-metadata columns in `output`. Today's code simply removes conflicting metadata attributes; this PR renames them instead. 

We could potentially explore removing that method entirely, but I don't know what side effects that might have?

Comment (ryan-johnson-databricks) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:368]: A quick local test of the change, with this query over a table with conflicting user schema:
```scala
df.select("_metadata")
  .withColumn("file_name", df.metadataColumn("_metadata").getField("file_name"))
```

... produces the expected logical plan (names conflict but `exprId` differs):
```
Project [_metadata#20, _metadata#24.file_name AS file_name#26]
+- Project [_metadata#20, _metadata#24]
   +- Relation [name#18,age#19,_METADATA#20,_metadata#24] parquet
```

... but query optimization fails with: 
```
[info]   java.lang.IllegalArgumentException: file_name does not exist. Available:
[info]   at org.apache.spark.sql.types.StructType.$anonfun$fieldIndex$1(StructType.scala:313)
[info]   at scala.collection.immutable.Map$EmptyMap$.getOrElse(Map.scala:110)
[info]   at org.apache.spark.sql.types.StructType.fieldIndex(StructType.scala:312)
  ...
[info]   at org.apache.spark.sql.catalyst.expressions.ProjectionOverSchema.getProjection(ProjectionOverSchema.scala:71)
  ...
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.$anonfun$buildNewProjection$3(SchemaPruning.scala:154)
  ...
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.buildNewProjection(SchemaPruning.scala:154)
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.org$apache$spark$sql$execution$datasources$SchemaPruning$$prunePhysicalColumns(SchemaPruning.scala:99)
  ...
[info]   at org.apache.spark.sql.execution.datasources.SchemaPruning$.apply(SchemaPruning.scala:41)
  ...
[info]   at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:211)
  ...
[info]   at org.apache.spark.sql.execution.QueryExecution.optimizedPlan$lzycompute(QueryExecution.scala:139)
  ...
[info]   at org.apache.spark.sql.QueryTest.checkAnswer(QueryTest.scala:148)
[info]   at org.apache.spark.sql.execution.datasources.FileMetadataStructSuite.$anonfun$new$13(FileMetadataStructSuite.scala:255)
```
So it appears that at least one optimizer rule fails to honor expression ids, and assumes the column names are unique?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:368]: @viirya this seems like a bug in nested column pruning

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:368]: Looks like `getProjection` wrongly matches attribute. It tries to get pruned datatype from `schema` by looking with field name. `StructType` maintains the mapping between (field name -> StructField), not (expr id -> StructField). If there are duplicate columns, it possibly picks incorrect attribute.

Comment (sigmod) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:368]: > If there are duplicate columns, it possibly picks incorrect attribute.

Is it possible to keep both in the requested schema?  

cc @cashmand

Comment (ryan-johnson-databricks) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:368]: I don't think there's anything in `StructType` that could track expr id (and IMO it shouldn't be trying to do so -- different levels of abstraction). 

It really looks like `StructType` doesn't allow duplicate names. The class is full of code that assumes uniqueness, often leveraging these private members: 
```scala
  private lazy val fieldNamesSet: Set[String] = fieldNames.toSet
  private lazy val nameToField: Map[String, StructField] = fields.map(f => f.name -> f).toMap
  private lazy val nameToIndex: Map[String, Int] = Utils.toMapWithIndex(fieldNames)
```
The `fieldNamesSet` goes all the way back to 2015...

**Comment by ryan-johnson-databricks** [comment]
> It's a good idea to provide an API that allows people to unambiguously reference metadata columns, and I like the new `Dataset.metadataColumn` function. However, I think the prepending underscore approach is a bit hacky. It's too implicit and I'd prefer a more explicit syntax like `SELECT metadata(_metadata) FROM t`. We can discuss this more and invite more SQL experts. Shall we exclude it from this PR for now?

@cloud-fan The prepended underscore is _NOT_ primarily intended as a user surface. Rather, it's a reliale way to get a unique column name that's still at least somewhat readable if you look at the query plan (unlike e.g. a uuid). The new `Dataset.metadataColumn` method does not even _look_ at a renamed attribute's name, for example.

At this point, the only reference in the code to prepended underscores is the two unit tests ("metadata name conflict resolved with leading underscores") that try to validate that the renaming works as intended. If you don't think the test coverage is important, we could remove even that?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 1 above)
Comment (ryan-johnson-databricks) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Removed `withMetadataColumn` method for now. Nothing stops us from adding it in the future if it's needed.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala (hunk 4 above)
Comment (ryan-johnson-databricks) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataStructSuite.scala:336]: Went ahead and added a test case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala (hunk 3 above)
Comment (olaky) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:56]: Comment is good, I just missed it

**Comment by cloud-fan** [comment]
about https://github.com/apache/spark/pull/40300/files#r1129818813 , I think if `SubqueryAlias` can't propagate metadata columns, then `df.metadataColumn` should not be able to get the column, what do you think? @ryan-johnson-databricks

File: sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala [hunk 6]
```diff
@@ -297,8 +297,14 @@ abstract class InMemoryBaseTable(
       InMemoryBatchScan(data.map(_.asInstanceOf[InputPartition]), schema, tableSchema)
 
     override def pruneColumns(requiredSchema: StructType): Unit = {
-      val schemaNames = metadataColumnNames ++ tableSchema.map(_.name)
-      schema = StructType(requiredSchema.filter(f => schemaNames.contains(f.name)))
+      // The required schema could contain conflict-renamed metadata columns, so we need to match
+      // them by their logical (original) names, not their current names.
+      val schemaNames = tableSchema.map(_.name).toSet
+      val prunedFields = requiredSchema.filter {
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:303]: Shall we move this handling to `V2ScanRelationPushDown`? in memory table is just a testing v2 source and we can't expect all existing v2 sources to make the same change like here.

[the remark this answers is not in the document]
Comment (ryan-johnson-databricks) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:303]: As far as I can tell, v2 intentionally puts the burden of pruning (and of metadata handling) on the datasource. So every datasource _does_ have to implement the policy it wants to use. 

But that's not necessarily a bad thing -- a datasource that chooses not to implement this change simply won't gain the ability to rename metadata columns on conflict.

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:303]: > ... simply won't gain the ability to rename metadata columns on conflict.

is it true? The rename happens at the Spark side, when Spark append metadata columns to the normal output columns.

[the remark this answers is not in the document]
Comment (ryan-johnson-databricks) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:303]: Today, if a dsv2 has a metadata column name conflict, the column is not selectable at all and attempts to do so will return the data column instead. The docs for `SupportsMetadataColumns` trait say:
>   * If a table column and a metadata column have the same name, the metadata column will never be
>   * requested and is ignored. It is recommended that Table implementations reject data column names
>   * that conflict with metadata column names.

After the changes in this PR:
* Implementations that follow the above advice won't need renaming in the first place
* Implementationt which do not follow the advice will continue to return the data column as before
* Users who attempt to call `df.metadataColumn` on a dsv2 that allows conflicts but doesn't support renaming yet would hit an analysis error.

[the remark this answers is not in the document]
Comment (ryan-johnson-databricks) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/connector/catalog/InMemoryBaseTable.scala:303]: Update: I made the renaming opt-in for dsv2 -- by default they fully keep existing behavior, but they can override `canRenameConflictingMetadataColumns` if they support automatic renaming.

**Comment by ryan-johnson-databricks** [comment]
> about https://github.com/apache/spark/pull/40300/files#r1129818813 , I think if `SubqueryAlias` can't propagate metadata columns, then `df.metadataColumn` should not be able to get the column, what do you think? @ryan-johnson-databricks

IMO changing the behavior of `SubqueryAlias` (and other specific plan node types) is out of scope for this PR -- this PR does not change that existing situation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala [hunk 7]
```diff
@@ -295,8 +297,9 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
             // Replace references to the _metadata column. This will affect references to the column
             // itself but also where fields from the metadata struct are used.
             case MetadataStructColumn(AttributeReference(_, fields @ StructType(_), _, _)) =>
-              CreateStruct(fields.map(
-                field => metadataColumns.find(attr => attr.name == newFieldName(field.name)).get))
+              val reboundFields = fields.map(
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:300]: unnecessary change?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsMetadataColumns.java [hunk 8]
```diff
@@ -48,11 +47,22 @@ public interface SupportsMetadataColumns extends Table {
...
    *
    * @return an array of {@link MetadataColumn}
    */
   MetadataColumn[] metadataColumns();
+
+  /**
+   * Determines how this data source handles name conflicts between metadata and data columns.
+   * <p>
+   * If true, spark will automatically rename the metadata column to resolve the conflict; metadata
+   * columns (renamed or not) can be reliably accessed by calling {@link Dataset.metadataColumn}.
```
[the remark this answers is not in the document]
Comment (ryan-johnson-databricks) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/SupportsMetadataColumns.java:61]: Updated to reference `MetadataAttributeWithLogicalName`, which hides the implementation details.

**Comment by cloud-fan** [comment]
thanks, merging to master!
